File: [第三方] 读写word演示脚本.mjs
/**
 * 小瓶RPA演示demo，具体api请查看*流程开发文档*
 * 官网：https://rpa.pbottle.com/
 * 流程开发文档：https://gitee.com/pbottle/pbottle-rpa/wikis/pages
 */

import pbottleRPA from "./pbottleRPA.js";  //必须含 .js 后缀
import fs from 'node:fs'
let mammoth
let docx
try {
	mammoth = await import('mammoth')  // import mammoth  from "mammoth"
    docx = await import('docx')   // import docx  from "docx"
} catch {
    pbottleRPA.showMsg('请先安装第三方模块','双击【第三方 模块安装.bat】')
    pbottleRPA.tts('请先安装第三方模块' + '双击【第三方 模块安装.bat】')
	pbottleRPA.exit('请先安装第三方模块' + '双击【第三方 模块安装.bat】')
}



console.log("=== word 后台读写测试 ===");
console.log(Date());
pbottleRPA.tts('word 后台读写测试')
pbottleRPA.sleep(1000*3)
pbottleRPA.tts(`将后台生成 word 文件`)
pbottleRPA.sleep(1000*5)

//生成 word 文档  更多例子：https://gitee.com/mirrors_dolanmiu/docx/tree/master/demo
const doc = new docx.Document({
    sections: [
        {
            properties: {},
            children: [
                new docx.Paragraph({
                    children: [
                        new docx.TextRun({
                            text: "标题文字",
                            bold: true,
                            size: 40,
                        }),
                    ],
                }),
                new docx.Paragraph({
                    children: [
                        new docx.TextRun({
                            text: "小瓶RPA官网：",
                        }),
                        new docx.ExternalHyperlink({
                            children: [
                                new docx.TextRun({
                                    text: "https://rpa.pbottle.com",
                                    style: "Hyperlink",
                                }),
                            ],
                            link: "https://rpa.pbottle.com",
                        }),
                    ],
                }),
            ],
        },
    ],
});

let buffer = await docx.Packer.toBuffer(doc);
fs.writeFileSync("word测试文档.docx", buffer);
pbottleRPA.openDir(pbottleRPA.__dirname)



//读取word文档
pbottleRPA.tts(`将后台读取 word 文件 显示到日志`)
pbottleRPA.sleep(1000*3)
let rs = await mammoth.extractRawText({path:"./word测试文档.docx"})
console.log('读取word文档内容：',rs.value);

File: python示例/[第三方] 读写word演示脚本.py
"""
PBottle RPA demo. Please refer to the *process documentation* for specific APIs.
Official website: https://rpa.pbottle.com/
Process documentation: https://rpa.pbottle.com/docs/
"""

import sys
import os
import time

# Add parent directory to path to import pbottleRPA
sys.path.insert(0, os.path.dirname(os.path.dirname(os.path.abspath(__file__))))
import pbottleRPA as rpa

# Try to import python‑docx and mammoth
try:
    from docx import Document
    from docx.shared import Pt
    from docx.enum.text import WD_ALIGN_PARAGRAPH
    import mammoth
except ImportError:
    rpa.showMsg(
        "Please install third‑party modules", "Run: pip install python-docx mammoth"
    )
    rpa.tts("Please install third‑party modules. Run: pip install python-docx mammoth")
    rpa.exit("Please install third‑party modules. Run: pip install python-docx mammoth")


print("=== Backend Word Read/Write Test ===")
print(rpa.getTime())
rpa.tts("Backend Word Read/Write Test")
rpa.wait(3)
rpa.tts("Generating a Word document in the background")
rpa.wait(5)

# Generate a Word document
doc = Document()

# Add a title paragraph
heading = doc.add_heading("Title Text", level=1)
heading.alignment = WD_ALIGN_PARAGRAPH.CENTER
run = heading.runs[0]
run.bold = True
run.font.size = Pt(20)

# Add a normal paragraph
p = doc.add_paragraph()
run1 = p.add_run("PBottle RPA official website: ")
run1.font.size = Pt(12)

# Add a hyperlink‑style text
hyperlink_run = p.add_run("https://rpa.pbottle.com")
hyperlink_run.font.size = Pt(12)
hyperlink_run.font.underline = True

# Add more sample content
doc.add_paragraph()
doc.add_paragraph("This is a Python‑based Word read/write demonstration.")
doc.add_paragraph("PBottle RPA supports a variety of automation operations, including:")
doc.add_paragraph("• Mouse and keyboard operations", style="List Bullet")
doc.add_paragraph("• Image recognition", style="List Bullet")
doc.add_paragraph("• AI capability integration", style="List Bullet")
doc.add_paragraph("• File operations", style="List Bullet")

# Save the file
word_path = os.path.join(
    os.path.dirname(os.path.dirname(os.path.abspath(__file__))),
    "Word_test_document.docx",
)
doc.save(word_path)
rpa.openDir(os.path.dirname(word_path))
rpa.tts("Word_test_document.docx has been generated. Please check.")
rpa.wait(3)

# Read the Word document
rpa.tts("Reading the Word document in the background and displaying content in the log")
rpa.wait(3)

# Use mammoth to extract text content
with open(word_path, "rb") as f:
    result = mammoth.extract_raw_text(f)
    print("Word document content:", result.value)

rpa.tts("Word test document has been read and logged.")
rpa.wait(2)


if __name__ == "__main__":
    pass
